File: actor/actorref.go
package actor

//ActorRef is an interface that defines the base contract for interaction of actors
type ActorRef interface {
	SendUserMessage(pid *PID, message interface{}, sender *PID)
	SendSystemMessage(pid *PID, message SystemMessage)
	Stop(pid *PID)
	Watch(pid *PID)
	UnWatch(pid *PID)
}

type LocalActorRef struct {
	mailbox Mailbox
}

func NewLocalActorRef(mailbox Mailbox) *LocalActorRef {
	return &LocalActorRef{
		mailbox: mailbox,
	}
}

func (ref *LocalActorRef) SendUserMessage(pid *PID, message interface{}, sender *PID) {
	if sender != nil {
		ref.mailbox.PostUserMessage(&UserMessage{Message: message, Sender: sender})
	} else {
		ref.mailbox.PostUserMessage(message)
	}
}

func (ref *LocalActorRef) SendSystemMessage(pid *PID, message SystemMessage) {
	ref.mailbox.PostSystemMessage(message)
}

func (ref *LocalActorRef) Stop(pid *PID) {
	ref.SendSystemMessage(pid, &Stop{})
}

func (ref *LocalActorRef) Watch(pid *PID) {
	ref.SendSystemMessage(pid, &Watch{Watcher: pid})
}

func (ref *LocalActorRef) UnWatch(pid *PID) {
	ref.SendSystemMessage(pid, &Unwatch{Watcher: pid})
}

func (ref *LocalActorRef) Suspend() {
	ref.mailbox.Suspend()
}

func (ref *LocalActorRef) Resume() {
	ref.mailbox.Resume()
}

File: languages/csharp/src/GAM/ActorRef.cs
﻿// //-----------------------------------------------------------------------
// // <copyright file="ActorRef.cs" company="Asynkron HB">
// //     Copyright (C) 2015-2016 Asynkron HB All rights reserved
// // </copyright>
// //-----------------------------------------------------------------------

using System.Threading.Tasks;

namespace GAM
{
    public class PID
    {
        public string Id { get; set; }

        public void Tell(object message)
        {
            var reff = ProcessRegistry.Instance.Get(this);
            reff.Tell(message);
        }

        public void SendSystemMessage(SystemMessage sys)
        {
            var reff = ProcessRegistry.Instance.Get(this);
            reff.SendSystemMessage(this, sys);
        }

        public void Request(object message, PID sender)
        {
            Tell(new Request(message, sender));
        }

        public async Task<T> RequestAsync<T>(object message)
        {
            var tsc = new TaskCompletionSource<T>();
            var p = Actor.FromProducer(() => new FutureActor<T>(tsc));
            var fpid = Actor.Spawn(p);
            Tell(new Request(message, fpid));
            await tsc.Task;
            return tsc.Task.Result;
        }

        public void Stop()
        {
            var reff = ProcessRegistry.Instance.Get(this);
            reff.Stop(this);
        }
    }

    public class Request
    {
        public Request(object message, PID sender)
        {
            Message = message;
            Sender = sender;
        }

        public object Message { get; }
        public PID Sender { get; }
    }

    public abstract class ActorRef
    {
        public abstract void Tell(object message);

        public void Stop(PID pid)
        {
            SendSystemMessage(pid, new Stop());
        }

        public abstract void SendSystemMessage(PID pid, SystemMessage sys);
    }

    public class LocalActorRef : ActorRef
    {
        public LocalActorRef(IMailbox mailbox)
        {
            Mailbox = mailbox;
        }

        public IMailbox Mailbox { get; }

        public override void Tell(object message)
        {
            Mailbox.PostUserMessage(message);
        }

        public override void SendSystemMessage(PID pid, SystemMessage sys)
        {
            Mailbox.PostSystemMessage(sys);
        }
    }
}